soc/xtensa/intel_adsp/tools/cavstool_client.py

```diff
@@ -41,27 +43,57 @@ def send_cmd(self, cmd):
 
             if ack == CMD_LOG_START:
                 self.monitor_log()
-            elif ack == CMD_LOG_STOP:
-                log.info(f"Stop output.")
             elif ack == CMD_DOWNLOAD:
                 self.run()
             else:
                 log.error(f"Receive incorrect msg:{ack} expect:{cmd}")
 
-    def download(self, filename):
+    def uploading(self, filename):
         # Send the FW to server
+        fname = os.path.basename(filename)
+        fsize = os.path.getsize(filename)
+        flen = len(fname)
+
+        md5_tx = hashlib.md5(open(filename,'rb').read()).hexdigest()
+
+        # Pack the header and the expecting packed size is 80 bytes.
+        # The header by convention includes:
+        # size(4), filename(42), MD5(32) and pad(2)
+        values = (fsize, fname.encode('utf-8'), md5_tx.encode('utf-8'), flen)
+        log.info(f'filename:{fname}, size:{fsize}, md5:{md5_tx}, len {flen}')
+
+        s = struct.Struct('I 42s 32s H')
+        packed_data = s.pack(*values)
+        packed_size = s.size
+        log.info(f'packed size: {packed_size}')
+
         with open(filename,'rb') as f:
-            log.info('Sending...')
-            ret = self.sock.sendfile(f)
-            log.info(f"Done Sending ({ret}).")
+            log.info(f'Sending...')
+
+            first_send = packed_data + f.read(BUF_SIZE - packed_size)
+            self.sock.send(first_send)
+            total = len(first_send)
+
+            while True:
+                l = f.read(1024)
```

Comment:
This seems to be re-implementing https://docs.python.org/3/library/os.html#os.sendfile and https://docs.python.org/3/library/socket.html#socket.socket.sendfile

This is Python, not C: every time you're writing basic code like this there is a 99.9% chance you're re-inventing some library function.

Reply:
Good suggestion! Simplify this part by using socket.send_file, thanks.